### PR TITLE
[FLINK-20491] Support Broadcast State in BATCH execution mode

### PR DESCRIPTION
**Pull request by aljoscha**
Right now, we don't support `DataStream.connect(BroadcastStream)` in `BATCH` execution mode. I believe we can add support for this with not too much work.

The key insight is that we can process the broadcast side before the non-broadcast side. Initially, we were shying away from this because of concerns about `ctx.applyToKeyedState()` which allows the broadcast side of the user function to access/iterate over state from the keyed side. We thought that we couldn't support this. However, since we know that we process the broadcast side first we know that the keyed side will always be empty when doing so. We can thus just make this "keyed iteration" call a no-op, instead of throwing an exception as we do now.

This is work-in-progress. I have yet to add tests.

## Brief change log

* turn broadcast operation into a logical `Transformation`
* add `BATCH` operators for broadcast
* add translators

## Verifying this change

To be done.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? maybe

We need to remove the caveat from the documentation. 🎊

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:349]: Do we need the `Concurrent` queue here? If I am not mistaken all of it happens from within the main mailbox thread.

**Comment by dawidwys** [comment]
I am good with the general idea and how it looks inside the `MultiInputSortingDataInput.java`. Unless someone objects/has a different opinion, I think we can start properly wiring all the inputs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:349]: Could be true, I went for a safe thing but could remove it. 👌

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/MultiInputTransformationTranslator.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/MultiInputTransformationTranslator.java:56]: I think this check is not sufficient now. We should check if any of the inputs has a state selector.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: Similar to the `MultiInput` shall we account for the situation if only one input is keyed?

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: I went with the current behaviour, but yes, we could do that. So in that case the non-keyed input would be processed before all the keyed input. Same for the general multi-input.

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: Yes, I'd do it that way. Otherwise I'd proceed with https://github.com/apache/flink/pull/14324. The current state leads to possibly corrupted results.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:48]: I'll do that then. 👌

**Comment by aljoscha** [comment]
I updated this with tests and more Javadocs. PTAL.

I still have to add support for multi-input operators and I need to enable mixed keyed/non-keyed operators, as per Dawid's suggestion.

**Comment by aljoscha** [comment]
@sjwiesman I'm also adding documentation here, it's a bit terse but still helpful, I think.

File: docs/dev/datastream_execution_mode.md
Comment (sjwiesman) [inline docs/dev/datastream_execution_mode.md:267]: @aljoscha What about this? 

```suggestion
### Order of Processing

The order in which records are processed in operators or user-defined functions (UDFs) can differ between `BATCH` and `STREAMING` execution.

In `STREAMING` mode, user-defined functions should not make any assumptions about incoming records' order.
Data is processed as soon as it arrives.

In `BATCH` execution mode, there are some operations where Flink guarantees order. 
The ordering can be a side effect of the particular task scheduling,
network shuffle, and state backend (see above), or a conscious choice by the system.

There are three general types of input that we can differentiate:

- _broadcast input_: input from a broadcast stream (see also [Broadcast
  State]({% link dev/stream/state/broadcast_state.md %}))
- _regular input_: input that isn't any of the above types of input
- _keyed input_: input from a `KeyedStream`

Functions, or Operators, that consume multiple input types will process them in the following order:

- broadcast inputs are processed first
- regular inputs are processed second
- keyed inputs are processed last

For functions that consume from multiple regular or broadcast inputs &mdash; such as a `CoProcessFunction` &mdash; Flink has the right to process data from any input of that type in any order. 
```

Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:267]: Perfect, thanks! I knew I was asking the right person.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:225]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:246]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:281]: Add `final` for uniformity.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:278]: Now the two `getBroadcastStateTransformation()` methods have no additional logic and each has a single call-site, so why not inlining them in their call-sites?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:51]: The two BroadcastTransformations could have a base class without the key-related fields and the function.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:124]: We can make the constructor `public` now and remove the static constructors from here and from the `keyed` alternative as now we have two separate classes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java:72]: I like that we create the operator here as now we push further down the runtime-related stuff. In the future we could have a translator that gives, for example, not an operator but sth else.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:62]: Now this can have `KS` as a generic type, right? (for type safety)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:64]: Same here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:140]: Now we can remove this and make the constructor `public`, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:764]: Why not putting it as a method in the `InputConfig` with a default implementation that returns `false`? I think that this will make the code nicer and also it will avoid checks like `inputConfig instanceof StreamConfig.NetworkInputConfig`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:234]: Can't we rename it to sth more descriptive?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:87]: Why not making this `final`?

Also nit-pick: I find it is nicer to initialize fields in the constructor but feel free to ignore this.

**Review by kl0u — CHANGES_REQUESTED** [review]
Thanks for the work @aljoscha . I had an initial look  and I left some comments. Unfortunately I am not done yet so I will continue with the review.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:764]: I decided to put it into the `NetworkInputConfig` because that's the only place where it can really be applied right now. But I can also move it to the base class.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:87]: None of the others are initialized in the constructor that's why I went with this one. But yes, I would agree!

**Comment by aljoscha** [comment]
Thanks for the review! I addressed most comments. I had to force push because the changes would have been a bit hard to follow if I had put them on top of the existing commits.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: This line seems to be undoing what you did before, right?

Comment (kl0u) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: Seems like a leftover line but that can lead to bugs.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:166]: Yes, it's a leftover. 🙈

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:115]: It is rather a matter of preference, but I'd move the index computing down below, where it is needed. I found it a bit confusing here and the IDE complains that in line 126 the `passThroughTaskInputs.size()` is always `0`.

We won't need this logic here, if we change the lines 151-160 to:
```StreamTaskInput<?>[] sortedInputs = selectableSortingInputs.getSortedInputs();
StreamTaskInput<?>[] passThroughInputs = selectableSortingInputs.getPassThroughInputs();
if (input1IsSorted) {
    input1 = toTypedInput(sortedInputs[0]);
} else {
    input1 = toTypedInput(passThroughInputs[0]);
}
if (input2IsSorted) {
    input2 = toTypedInput(sortedInputs[sortedInputs.length - 1]);
} else {
    input2 = toTypedInput(passThroughInputs[passThroughInputs.length - 1]);
}
```

We can base that on the assumption that the `input1` will be first in the corresponding group while the `input2` last.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:130]: nit: my IDE complains the `"rawtypes"` is unnecessary.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I think we should change it so that we apply sorting if any of the inputs is keyed.

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:196]: nit: How about we test with at least two keys?

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thanks a lot for the great work!

A general comment (that I also put inline), I'd be interested to see the impact of the change on benchmarks results.

File: docs/dev/datastream_execution_mode.md
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:255]: regular input should be processed before keyed input, because the (sorted) keyed input is always last.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:263]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java:36]: It's not correct and I copied it from the previous transformation. I can remove the sentence or try and fix it. WDYT?

File: docs/dev/datastream_execution_mode.md
Comment (dawidwys) [reply docs/dev/datastream_execution_mode.md:255]: I think I might've not been clear on this one. What I meant is that the line with `regular input` should be below the `keyed input`, because it refers to it. If I understood the sentence correctly it says: `input that isn't broadcast or keyed`, right?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: It's technically wrong but we don't allow mixed sorted/non-sorted inputs right now, I think.

In general, the support for mixed inputs is in flux with this PR, also relevant for https://github.com/apache/flink/pull/14324.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java:36]: I'd be ok with simply dropping it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:115]: fixing

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:130]: fixed

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: I don't think there is anything that prohibits mixing keyed with non-keyed inputs. Users can manually craft such Transformation. As for the #14324 didn't we say we will drop it once we merge this PR with a support for mixing such inputs?

BTW, If I remember correctly we had that discussion already ;)

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java
Comment (aljoscha) [reply flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:196]: adding it to the test 👌

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I think we have to yes. I shied away from it here because I would have to add a test that uses the low-level API because right now it's not possible for users to create such transforms.

Could you maybe whip together a test based on the existing test code you already had? We add it here and I change it to correctly forward the requirements?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: As for the #14324 if I remember correctly we said we will drop it and support mixed keyed/non-keyed inputs: https://github.com/apache/flink/pull/14312#discussion_r540209912

As for the multi inputs those are not exposed in a nice API at all, but I don't think there are any checks or anything that would prohibit users from mixing keyed with non-keyed inputs.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:353]: changing to a `LinkedList` 👌

File: docs/dev/datastream_execution_mode.md
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:255]: Ah damn, you're right... 🙈 fixing it

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: sure, I will prepare one

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: I was thinking of tests like those: https://github.com/dawidwys/flink/blob/mix-keyed-2/flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:49]: Sorry, I also got started on it and should have notified you. 🙈

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:114]: dropping it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java]: dropping it

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:114]: results are now on http://codespeed.dak8s.net:8000/comparison/

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:151]: I added support for mixed two-input operations, along with tests. I found some bugs along the way, the fixes are individual commits to make it easier to follow/review.

File: docs/dev/datastream_execution_mode.md
Comment (dawidwys) [inline docs/dev/datastream_execution_mode.md:3]: Are all the changes in the file required? Moreover I think this change to the header may break the page.

**Comment by dawidwys** [comment]
What is the plan for mixed keyed/non-keyed inputs in multiple inputs? If I am not mistaken the implementation in the `StreamMultipleInputProcessorFactory` is not complete and it does not set the pass through inputs correctly.

File: docs/dev/datastream_execution_mode.md
Comment (aljoscha) [reply docs/dev/datastream_execution_mode.md:3]: I don't know how those snuck in... 🙈, reverting

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:538]: Actually the non-broadcasted input is keyed, so we should probably remove the comment.

I copied it over from one of your tests and forgot to remove the comment. The same applies for the other manually crafted two input broadcast operator.

**Review by dawidwys — APPROVED** [review]
+1 % Azure

Looks good now!

**Comment by aljoscha** [comment]
Merged this manually

**Pull request closed without merging**
